linera-storage/src/lib.rs

```diff
@@ -366,6 +366,7 @@ pub struct ChainRuntimeContext<S> {
     execution_runtime_config: ExecutionRuntimeConfig,
     user_contracts: Arc<DashMap<UserApplicationId, UserContractCode>>,
     user_services: Arc<DashMap<UserApplicationId, UserServiceCode>>,
+    blobs: Arc<DashMap<BlobId, Blob>>,
```

Comment:
`ChainRuntimeContext`s can now remain in memory for a long time, and for any number of blocks. For the contracts and services it makes sense to keep them in memory, because an application that has been used on a chain (e.g. `non-fungible`) it's likely that it will be used again.
But for other—potentially large—blobs (e.g. NFT data!) it's often the case that they get only used once (maybe the NFT is traded on this chain and moves elsewhere). In that case, we shouldn't keep the blob in memory indefinitely.

Reply:
Good point! Will replace it with a `ValueCache`

Reply:
But is it really necessary at all?

For the `chain_tests` above: Can't we just write the blobs to "storage", i.e. to the `MemoryStore` that the test is using?

And in production: We're currently adding blob oracle responses too often, probably. In particular for bytecode blobs, would it be sufficient to only add them the _first_ time they are used on any chain, i.e. whenever we put the contracts and services in the two maps above?

Then for all the future blocks, we don't need to check for the presence of the blobs anymore, and the bytecode is already loaded in those maps?

Reply:
The thing is that for every block we try to execute, we need to know what blobs are "required" in the execution of that block. And they need to either be present in storage or in some pending state as this is the block publishing it. So I think to know somehow these required blobs, and right now we use the oracle responses for that 🤔
Is there a better way of doing this? If we just assume that after publishing they're there then we'll waste more time before failing, so IMHO it's better to continue to fail early like we do now, but not sure if you had other ideas
Regarding the blobs in the `ChainRuntimeContext`, I can see if it can work without it

Reply:
Talked offline, if we're not in the block that publishes the blob, we can check if the execution state has an application state for this application. If it does, we don't need to check for that application or the bytecode blobs for it. Otherwise we continue checking as we do now. I will do that in a separate PR though, for this one I'll just see if this map can be removed.

Reply:
https://github.com/linera-io/linera-protocol/issues/2620